proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,440 @@
+to write pun-free code. In addition, add a new extension that disables
+the built-in syntax for these types, thus requiring users to use the new
+names.
+
+.. _`#270`: https://github.com/ghc-proposals/ghc-proposals/pull/270
+.. _`#281`: https://github.com/ghc-proposals/ghc-proposals/pull/281
+.. _`#378`: https://github.com/ghc-proposals/ghc-proposals/pull/378
+.. _`#473`: https://github.com/ghc-proposals/ghc-proposals/pull/473
+
+Motivation
+----------
+Proposals `#270`_, `#281`_, `#378`_, and `#473`_ all struggle when faced
+with *puns*, the use of an identifier that can be interpreted either in the
+type-level namespace or the term-level namespace (but with different meanings).
+
+Furthermore, puns can be hard for newcomers to Haskell. It is not hard to jump
+from ::
+
+  single :: [Int]
+  single = [5]
+
+to ::
+
+  just :: Maybe Int
+  just = Maybe 5
+
+even though the latter is nonsense. With this proposal, users would have the
+option of writing ::
+
+  single :: List Int
+  single = [5]
+
+Puns also cause trouble when describing what we are talking about: conversations
+around these features invariably require people to say "``[]``-the-type" or
+"``[]``-the-empty-list" or similar.
+
+This proposal thus allows for a common way to refer to tuples and lists without
+punning. It is entirely opt-in, though we expect the new phrasing to become popular
+over time.
+
+Proposed Change Specification
+-----------------------------
+
+1. Create a new module in ``base`` called ``GHC.Prelude``. This module exports definitions
+   peculiar to GHC that are safe to use in ordinary code.
+   This module is *not* intended as a replacement to ``Prelude``,
+   but instead a complement to it. This module is a counterpart to ``GHC.Exts``, which exports
+   many *unsafe* internals.
+
+   This proposal lists several definitions to be exported from ``GHC.Prelude``, but leaves other
+   definitions to be added in separate proposal(s).
+
+#. Export the following definitions from ``GHC.Tuple`` and ``GHC.Prelude``. (Implementation note:
+   These would probably end up in a new module ``GHC.Tuple.Prim``, due to the dependency from
+   the definitions below in ``GHC.Tuple`` on ``GHC.TypeLits``.) ::
+
+     data Unit = ()
+     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`
+
+     {-# DEPRECATED Solo "The Solo constructor has been renamed to MkSolo to avoid punning." #-}
+     pattern Solo :: a -> Solo a
+     pattern Solo x = MkSolo x
+     {-# COMPLETE Solo #-}
+
+     getSolo :: Solo a -> a
+     getSolo (Solo a) = a   -- as today
+
+     type Tuple0 = Unit
+     type Tuple1 = Solo
+     data Tuple2 a b = (a, b)
+     data Tuple3 a b c = (a, b, c)
+     -- ...
+     data Tuple64 ... = (...)
+
+#. Export the following definitions from ``GHC.Exts`` and ``GHC.Prelude``. These
+   replace similar definitions (in ``GHC.Classes``) today. (Note that ``(...) =>`` is special syntax, and does not
```

Comment:
This still sounded a bit mysterious to me. So, for anybody who is not in the know, today (as of Long Ago™), `GHC.Classes` hosts the following class definitions (which are not exported, hence not present in the Haddock). It would be helpful to any future person reading the text if this was explained in the proposal.

```haskell
class ()                                                                                                                                                                                                                                      
class (c1, c2)     => (c1, c2)                                                                                                                                                                                                                
class (c1, c2, c3) => (c1, c2, c3)                                                                                                                                                                                                            
class (c1, c2, c3, c4) => (c1, c2, c3, c4)                                                                                                                                                                                                    
class (c1, c2, c3, c4, c5) => (c1, c2, c3, c4, c5)                                                                                                                                                                                            
class (c1, c2, c3, c4, c5, c6) => (c1, c2, c3, c4, c5, c6)                                                                                                                                                                                    
class (c1, c2, c3, c4, c5, c6, c7) => (c1, c2, c3, c4, c5, c6, c7) 
…
class (c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16,                                                                                                                                                                 
       c17, c18, c19, c20, c21, c22, c23, c24, c25, c26, c27, c28, c29, c30,                                                                                                                                                                  
       c31, c32, c33, c34, c35, c36, c37, c38, c39, c40, c41, c42, c43, c44,                                                                                                                                                                  
       c45, c46, c47, c48, c49, c50, c51, c52, c53, c54, c55, c56, c57, c58,                                                                                                                                                                  
       c59, c60, c61, c62, c63, c64)                                                                                                                                                                                                          
   => (c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16,                                                                                                                                                                 
       c17, c18, c19, c20, c21, c22, c23, c24, c25, c26, c27, c28, c29, c30,                                                                                                                                                                  
       c31, c32, c33, c34, c35, c36, c37, c38, c39, c40, c41, c42, c43, c44,                                                                                                                                                                  
       c45, c46, c47, c48, c49, c50, c51, c52, c53, c54, c55, c56, c57, c58,                                                                                                                                                                  
       c59, c60, c61, c62, c63, c64) 
```

Reply:
I've tried to clarify this.